diff --git a/cpu-exec.c b/cpu-exec.c
index 38e5f02..283e8c7 100644
--- a/cpu-exec.c
+++ b/cpu-exec.c
@@ -23,6 +23,13 @@
 #include "qemu/atomic.h"
 #include "sysemu/qtest.h"
 
+#include <unistd.h>
+#include "shared.h"
+#include <signal.h>
+
+int stop_for_map = 0;
+char pause_addr[128] = "";
+
 void cpu_loop_exit(CPUState *cpu)
 {
     cpu->current_tb = NULL;
@@ -620,6 +627,13 @@ int cpu_exec(CPUArchState *env)
                     qemu_log("Trace %p [" TARGET_FMT_lx "] %s\n",
                              tb->tc_ptr, tb->pc, lookup_symbol(tb->pc));
                 }
+
+		uint32_t converted_addr = (uint32_t)strtol(pause_addr, NULL, 0);
+                if (stop_for_map == 1) {
+                    if (converted_addr == tb->pc) {
+                        pause();
+                    }
+                }
                 /* see if we can patch the calling TB. When the TB
                    spans two pages, we cannot safely do a direct
                    jump. */
diff --git a/linux-user/main.c b/linux-user/main.c
index b453a39..472a42c 100644
--- a/linux-user/main.c
+++ b/linux-user/main.c
@@ -34,6 +34,8 @@
 #include "qemu/envlist.h"
 #include "elf.h"
 
+#include "shared.h"
+
 char *exec_path;
 
 int singlestep;
@@ -3472,6 +3474,12 @@ static void handle_arg_help(const char *arg)
     usage();
 }
 
+static void handle_arg_pause(const char *arg)
+{
+    stop_for_map = 1;
+    strncpy(pause_addr, arg, 128);
+}
+
 static void handle_arg_log(const char *arg)
 {
     int mask;
@@ -3643,6 +3651,8 @@ struct qemu_argument {
 static const struct qemu_argument arg_table[] = {
     {"h",          "",                 false, handle_arg_help,
      "",           "print this help"},
+    {"b",          "",                 true,  handle_arg_pause,
+     "address",    "pause at addr for memory map"},
     {"g",          "QEMU_GDB",         true,  handle_arg_gdb,
      "port",       "wait gdb connection to 'port'"},
     {"L",          "QEMU_LD_PREFIX",   true,  handle_arg_ld_prefix,
diff --git a/shared.h b/shared.h
new file mode 100644
index 0000000..860f7ca
--- /dev/null
+++ b/shared.h
@@ -0,0 +1,2 @@
+extern int stop_for_map;
+extern char pause_addr[128];
